crates/ty_python_semantic/src/types/builder.rs

```diff
@@ -444,93 +444,96 @@ impl<'db> UnionBuilder<'db> {
                     .filter_map(UnionElement::to_type_element)
                     .any(|ty| Type::EnumLiteral(enum_member_to_add).is_subtype_of(self.db, ty))
                 {
-                    self.elements
-                        .push(UnionElement::Type(Type::EnumLiteral(enum_member_to_add)));
```

Comment:
This was the problem. Instead of just using `self.elements.push`, we should go through the whole simplification logic in the `_ => {…}` branch, which I have now moved to a separate function. Every other change in this file is just the "moving out to a separate function" part.